### ISSUE TITLE
Accurarcy() does not work, but 'accuracy' does

### ISSUE THREAD

**Issue by LostInDarkMath** [keras-team/tf-keras#15]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): not really
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 22.04
- TensorFlow installed from (source or binary): `pip install tensorflow`
- TensorFlow version (use command below): `2.13.0`
- Python version: `3.11.5`
- Bazel version (if compiling from source): - 
- GPU model and memory: does not matter
- Exact command to reproduce: run script below

**Describe the problem**.
If I write `Accurarcy()` in the `metrics` list, it does not work. But the String `accuracy` does work. According to the docs, bith sould work. See example code below.

**Describe the current behavior**.
```
469/469 [==============================] - 3s 6ms/step - loss: 0.2548 - accuracy: 0.0000e+00
```

**Describe the expected behavior**.
```
469/469 [==============================] - 3s 6ms/step - loss: 0.2540 - accuracy: 0.9260
```

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): no
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing): - 

**Standalone code to reproduce the issue**.

```python
import numpy as np
from keras import Sequential
from keras.datasets import mnist
from keras.src import activations
from keras.src.layers import Dense
from keras.src.losses import CategoricalCrossentropy
from keras.src.metrics import Accuracy
from keras.src.optimizers import RMSprop
from keras.src.utils import to_categorical


def preprocess_images(images):
    s = images.shape
    return images.reshape((s[0], s[1] * s[2])).astype(dtype=np.float32) / 255


if __name__ == '__main__':
    (train_images, train_labels), (test_images, test_labels) = mnist.load_data()
    processed_train_images = preprocess_images(train_images)
    processed_test_images = preprocess_images(test_images)
    processed_train_labels = to_categorical(y=train_labels)
    processed_test_labels = to_categorical(y=test_labels)

    network = Sequential()
    network.add(layer=Dense(units=512, activation=activations.relu, input_shape=(28 * 28, )))
    network.add(layer=Dense(units=10, activation=activations.softmax))

    network.compile(optimizer=RMSprop(), loss=CategoricalCrossentropy(), metrics=[Accuracy()])  # this does not work
    # network.compile(optimizer=RMSprop(), loss=CategoricalCrossentropy(), metrics=['accuracy'])  # this works

    network.fit(x=processed_train_images, y=processed_train_labels, epochs=1, batch_size=128)
```

**Source code / logs**.
Nothing.

**Comment by tilakrayal** [comment]
@LostInDarkMath,
Apologies for the delay. Here in the code you are trying to use `loss=CategoricalCrossentropy()`, where Accuracy is not the better option. Instead you can use **keras.metrics.CategoricalAccuracy** for the accurate result. I tried to execute with keras.metrics.CategoricalAccuracy and it provided the required output. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/bbbc27edcea78332b3e53d1b3ac5b6f8/untitled1390.ipynb).

![Screenshot 2023-10-16 2 53 55 PM](https://github.com/keras-team/tf-keras/assets/81610181/3d9c58fc-0f22-461f-884b-1fb2d0f2462c)


https://www.tensorflow.org/api_docs/python/tf/keras/metrics/CategoricalAccuracy
Thank you!

**Comment by LostInDarkMath** [comment]
Thank your for the clarification.

That means that the string `'accuracy'` is resolved differently depending on the loss function? Or does `'accuracy'` always resolve to `eras.metrics.CategoricalAccuracy`? And where is dthis behavior documented? Is there something like a mapping table that visualizes this mapping? Would be very nice :)

**Comment by innat** [comment]
@LostInDarkMath 
Here the following API are technically same. 

```python
keras.metrics.Accuracy
keras.metrics.BinaryAccuracy
```

From [doc](https://keras.io/api/metrics/accuracy_metrics/#accuracy-class), 

> keras.metrics.Accuracy(name="accuracy", ...)
> keras.metrics.BinaryAccuracy(name="binary_accuracy", ...)
Calculates how often predictions equal labels.
This metric creates two local variables, total and count that are used to compute the frequency with which y_pred matches y_true. This frequency is ultimately returned as **binary accuracy**: an idempotent operation that simply divides total by count.

And when we use string identifier, for example [`accuracy`], later it converts to the appropriate metrics based on the labels and logits, [source](https://github.com/keras-team/keras/blob/68f9af408a1734704746f7e6fa9cfede0d6879d8/keras/engine/training.py#L684-L689). 

> When you pass the
              strings 'accuracy' or 'acc', we convert this to one of
              `tf.keras.metrics.BinaryAccuracy`,
              `tf.keras.metrics.CategoricalAccuracy`,
              `tf.keras.metrics.SparseCategoricalAccuracy` based on the shapes
              of the targets and of the model output. We do a similar


@tilakrayal 
Apart from the abvove clarification, there may be potential bug. Please check this [question](https://stackoverflow.com/q/77534387/9215780) on stack overflow. That is very similar to this issue.

**Comment by sachinprasadhs** [comment]
From the model.copile document:

> List of metrics to be evaluated by the model during training and testing. Each of this can be a string (name of a built-in function), function or a [tf.keras.metrics.Metric](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/Metric) instance. See [tf.keras.metrics](https://www.tensorflow.org/api_docs/python/tf/keras/metrics). Typically you will use metrics=['accuracy']. A function is any callable with the signature result = fn(y_true,y_pred). To specify different metrics for different outputs of a multi-output model, you could also pass a dictionary, such as metrics={'output_a':'accuracy', 'output_b':['accuracy', 'mse']}. You can also pass a list to specify a metric or a list of metrics for each output, such as metrics=[['accuracy'], ['accuracy', 'mse']] or metrics=['accuracy', ['accuracy', 'mse']]. When you pass the strings 'accuracy' or 'acc', we convert this to one of [tf.keras.metrics.BinaryAccuracy](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/BinaryAccuracy), [tf.keras.metrics.CategoricalAccuracy](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/CategoricalAccuracy), [tf.keras.metrics.SparseCategoricalAccuracy](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/SparseCategoricalAccuracy) based on the shapes of the targets and of the model output. We do a similar conversion for the strings 'crossentropy' and 'ce' as well. The metrics passed here are evaluated without sample weighting; if you would like sample weighting to apply, you can specify your metrics via the weighted_metrics argument instead.

**Comment by sampathweb** [comment]
When you use [metrics.Accuracy](https://github.com/keras-team/tf-keras/blob/master/tf_keras/metrics/accuracy_metrics.py#L29), it resolves to `BinaryAccuracy` as stated in the doc of `Accuracy`.  Whereas, when using string `accuracy` it resolves to BinaryAccuracy or CategoricalAccuracy, based on the target shape.  This was also explained by @innat above. This may be confusing, so we recommend either explicitly using `BinaryAccuracy` or `CategoricalAccuracy` if you don't want to use the string `accuracy` for the metrics.

**Comment by sampathweb** [comment]
The same behavior exists in Keras 3 as well.  Should we deprecate use of `metrics.Accuracy` or mimic its usage to string metric of `accuracy`?

**Comment by innat** [comment]
> The same behavior exists in Keras 3 as well.  Should we deprecate use of `metrics.Accuracy` or mimic its usage to string metric of `accuracy`?

The `metrics.Accuracy` can be removed. It makes confusion, especially among the beginners a lot.

**Comment by amitsrivastava78** [comment]
Thanks for reporting this! You're absolutely right that there's confusing behavior here. Let me explain what's happening and provide the solution.
The Issue
metrics.Accuracy() performs exact equality comparison between y_true and y_pred. For categorical classification with one-hot encoded labels and softmax outputs, this will always give 0 accuracy because:
* y_true: [0, 0, 1, 0] (one-hot encoded)
* y_pred: [0.1, 0.2, 0.6, 0.1] (softmax probabilities)
These will never be exactly equal, so accuracy is always 0.
The Solution
For categorical classification, you should use CategoricalAccuracy instead:
```
# ❌ This doesn't work correctly
network.compile(metrics=[Accuracy()])

# ✅ Use one of these instead:
network.compile(metrics=['accuracy'])  # Auto-resolves to CategoricalAccuracy
network.compile(metrics=[CategoricalAccuracy()])  # Explicit
```

Why This Happens
The string 'accuracy' gets automatically converted to the appropriate accuracy metric based on your loss function and tensor shapes:
* CategoricalCrossentropy → CategoricalAccuracy
* BinaryCrossentropy → BinaryAccuracy
* SparseCategoricalCrossentropy → SparseCategoricalAccuracy
Ongoing Discussion
The Keras team is currently discussing whether to deprecate metrics.Accuracy entirely since it's confusing and rarely what users actually want. The string shortcut 'accuracy' works much more intuitively.
For now, I'd recommend always using the string 'accuracy' or the explicit metric classes (CategoricalAccuracy, BinaryAccuracy, etc.) to avoid this confusion.
